numba/targets/arraymath.py

```diff
@@ -1380,6 +1380,81 @@ def np_vander_seq_impl(x, N=None, increasing=False):
     elif isinstance(x, (types.Tuple, types.Sequence)):
         return np_vander_seq_impl
 
+#----------------------------------------------------------------------------
+# Mathematical functions
+
+@register_jitable
+def np_interp_impl_inner(x, xp, fp, out_dtype):
+    x_arr = np.asarray(x)
+    xp_arr = np.asarray(xp)
+    fp_arr = np.asarray(fp)
+
+    if len(xp_arr) == 0:
+        raise ValueError('array of sample points is empty')
+
+    if len(xp_arr) != len(fp_arr):
+        raise ValueError('fp and xp are not of the same length.')
+
+    if xp_arr.size > 1 and not np.all(np.diff(xp_arr) > 0):
+        raise ValueError('The x-coordinates of the data points must be increasing')
+        # this is a strict and literal interpretation of NumPy docs:
+        # https://github.com/numpy/numpy/blob/v1.15.0/numpy/lib/function_base.py#L1192
+        #   "The x-coordinates of the data points, must be increasing if argument
+        #   `period` is not specified."
+        # NumPy does not check for this or enforce it; whereas NumPy will return output,
+        # this implementation will raise.
+
+    if xp_arr.size == 1:
+        return np.full(x_arr.shape, fill_value=fp_arr[0], dtype=out_dtype)
+    else:
+        out = np.empty(x_arr.shape, dtype=out_dtype)
+
+        for i in range(x_arr.size):
+            if x_arr.flat[i] > xp_arr[-1]:
+                out.flat[i] = fp_arr[-1]
+            elif x_arr.flat[i] < xp_arr[0]:
+                out.flat[i] = fp_arr[0]
+            else:
+                idx = np.searchsorted(xp_arr, x_arr.flat[i])
+                f = (x_arr.flat[i] - xp_arr[idx - 1]) / (xp_arr[idx] - xp_arr[idx - 1])
+                out.flat[i] = fp_arr[idx - 1] + f * (fp_arr[idx] - fp_arr[idx - 1])
+
+        return out
+
+if numpy_version >= (1, 10):
+    # replicate behaviour of NumPy 1.10 where handling of
+    # unsorted xp was changed
+    @overload(np.interp)
+    def np_interp(x, xp, fp):
+
+        for param in xp, fp:
+            if hasattr(param, 'ndim') and param.ndim > 1:
+                raise TypingError('xp and fp must both be 1D')
+
+        complex_dtype_msg = (
+            "Cannot cast array data from complex dtype to float64 dtype"
+        )
+
+        if hasattr(xp, 'dtype'):
+            if np.issubdtype(as_dtype(xp.dtype), np.dtype(complex).type):
```

Comment:
Is this like `np.iscomplexobj` ?

Reply:
Oh, yeah I think so

Reply:
Changed this (but I'm in the typing layer, so still a test of dtype rather than object, if you see what I mean).